content/src/content/docs/learn/tutorials/hooking-up-a-database.mdx

```diff
@@ -0,0 +1,395 @@
+import { createServer } from "node:http"
+import { Client } from "pg";
+const db = new Client({ user: "postgres" });
+const health = Effect.sync(() => {
+  return HttpServerResponse.text("ok")
+})
+//---cut---
+// @errors: 2379
+const users = Effect.promise(async () => {
+  const { rows } = await db.query("SELECT * FROM users");
+  return HttpServerResponse.json({ users: rows });
+});
+
+const router = HttpRouter.empty.pipe(
+  HttpRouter.get("/health", health),
+  HttpRouter.get("/users", users),
+)
+```
+
+This doesn't work, as we can see from the error. The reason is that
+`HttpServerResponse.json` is an operation that might fail. Not all JavaScript
+objects can be converted to JSON.
+
+```ts twoslash
+JSON.stringify({ big: 10n })
+//=> TypeError: JSON.stringify cannot serialize BigInt.
+```
+
+For this reason, `HttpServerResponse.json` returns an `Effect` that can fail.
+We didn't see this earlier because `HttpServerResponse.text` can't fail, and
+so just returns an `HttpServerResponse` directly.
+
+## Error handling in Effect
+
+When we introduced the `Effect` type in the previous section, we dealt
+exclusively with Effects that couldn't fail. That won't get us very far if
+we're building anything non-trivial.
+
+`Effect` is a generic type that takes 3 arguments:
+
+```typescript
+Effect<Success, Error, Requirements>
+```
+
+`Success` is the type the Effect returns when it succeeds, `Error` is the type
+it returns when it fails, and we'll talk about `Requirements` a bit later.
+
+When we created our example Effects in the previous section, these all took on
+the type `Effect<string, never, never>`.
+
+```ts twoslash
+import { Effect } from "effect"
+//---cut---
+Effect.sync(() => "Hello, world!")
+//=> Effect.Effect<string, never, never>
+```
+
+This means they will return a `string` on success, and never fail.
+
+The Effect that `HttpServerResponse.json` returns is of type
+`Effect<HttpServerResponse, HttpBodyError, never>`. So when we try to return
+it from our `users` handler, we end up wrapping one Effect in another.
+
+```ts twoslash
+import { HttpServerResponse } from "@effect/platform"
+import { Effect, Layer } from "effect"
+import { createServer } from "node:http"
+import { Client } from "pg";
+const db = new Client({ user: "postgres" });
+//---cut---
+const users = Effect.promise(async () => {
+//    ^?
+  const { rows } = await db.query("SELECT * FROM users");
+  return HttpServerResponse.json({ users: rows });
+});
+```
+
+This is what the type error from earlier was trying to tell us. Here's the
+error again:
+
+```text wrap
+Argument of type 'Effect<Effect<HttpServerResponse, HttpBodyError, never>, never, never>' is not assignable to parameter of type 'Handler<never, never>'
+```
+
+Under the hood, a `Handler<never, never>` is an `Effect<HttpServerResponse,
+never, never>`. Because `Effect<HttpServerResponse, HttpBodyError, never>` is
+not an `HttpServerResponse`, the type checker gets upset with us.
```

Comment:
```suggestion
Under the hood, a `Handler<never, never>` is just a type alias for `Effect<HttpServerResponse,
never, never>`. Because `Effect<HttpServerResponse, HttpBodyError, never>` is
not an `HttpServerResponse`, the type checker gets upset with us.
```